terminal.go

```diff
@@ -29,29 +34,42 @@ type Terminal struct {
 // to restore the terminal to its original state upon exit.
 // fortio.org/log (and thus stdlib "log") will be redirected
 // to the terminal in a manner that preserves the prompt.
-func Open() (*Terminal, error) {
+// New cancellable context is returned, use it to cancel the terminal
+// reading or check for done for control-c or signal.
+func Open(ctx context.Context) (t *Terminal, err error) {
+	intrReader := NewInterruptReader(os.Stdin, 256)
 	rw := struct {
 		io.Reader
 		io.Writer
-	}{os.Stdin, os.Stderr}
-	t := &Terminal{
-		fd: int(os.Stdin.Fd()),
+	}{intrReader, os.Stderr}
+	t = &Terminal{
+		fd:         int(os.Stdin.Fd()), //nolint:gosec // yeah right.
+		intrReader: intrReader,
+		Context:    ctx,
 	}
 	t.term = term.NewTerminal(rw, "")
 	t.Out = t.term
 	if !t.IsTerminal() {
 		t.Out = os.Stderr // no need to add \r for non raw mode.
-		return t, nil
+		t.ResetInterrupts(ctx)
+		return
 	}
-	var err error
 	t.oldState, err = term.MakeRaw(t.fd)
 	if err != nil {
-		return nil, err
+		return
 	}
 	t.term.SetBracketedPasteMode(true) // Seems useful to have it on by default.
 	t.capacity = term.DefaultHistoryEntries
 	t.loggerSetup()
-	return t, nil
+	t.ResetInterrupts(ctx)
+	return
+}
+
+// If you want to reset and restart after an interrupt, call this.
+func (t *Terminal) ResetInterrupts(ctx context.Context) (context.Context, context.CancelFunc) {
+	// locking should not be needed as we're (supposed to be) in the main thread.
+	t.Context, t.Cancel = t.intrReader.Start(ctx)
+	return t.Context, t.Cancel
```

Comment:
Unless I'm wrong, you are never using the returned values

Reply:
I added demo use now, but it'll also be used in grol, at least the context if not the cancel function

Reply:
Understood